api/system/src/lib.rs

```diff
@@ -247,10 +248,25 @@ impl<Api: api::Api> System<Api> {
 		let _ = dt; // this to prevent earlier drop.
 		epoch
 	}
-}
 
+	/// Equivalent to [`sys::ffi::playdate_sys::setSerialMessageCallback`]
+	#[doc(alias = "sys::ffi::playdate_sys::setSerialMessageCallback")]
+	// NOTE: ideally this would be similar to our implementation of `fs::read_dir`
+	//
+	// i.e.
+	//
+	//   pub fn set_serial_message_callback<Fn>(&self, callback: Fn) where Fn: FnMut(String) {
+	//
+	// but we can't do that because of
+	// https://devforum.play.date/t/feature-request-add-a-userdata-argument-to-setserialmessagecallback/17333
+	pub fn set_serial_message_callback(&self, callback: Option<unsafe extern "C" fn(data: *const c_char)>) {
+		let f = self.0.set_serial_message_callback();
+		unsafe { f(callback) }
+	}
+}
```

Comment:
I agree that having userdata would be really helpful. But it is __not critically__ necessary. You can do without it, albeit with a little overhead.

We still can use [typed storage][storage], that's compiletime kinda magic. If explain shortly, any `Fn` type in the two levels of AST in the rustc have unique type, so we can 
- store it with unique key
- store type as type-param (generic)
- get it back with type-param 

\* _i mainly mean closures, not fn-pointers_

For example see my implementation in scoreboards - [storage][] and [usage][].

_This comment — it's just a comment and a suggestion for discussion, not an urgent demand for change, of course._ 😬

[storage]: https://github.com/boozook/playdate/blob/main/api/scoreboards/src/storage.rs
[usage]: https://github.com/boozook/playdate/blob/main/api/scoreboards/src/lib.rs#L83-L85

Reply:
Also, I don't see any reason for Dave to add userdata for that callback. On one side, hi stores just one ptr, but on other side hi have to store ptr and mapping like ptr->userdata. I'm sure he'll say it's an overhead and won't do it without more reasons. 🤷🏻‍♂️

Reply:
Thanks! Please see c6ec4474a39c4ddd443abf1b37d4181f7d07f69f